pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -744,6 +744,19 @@ func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus)
 		return nil, nil, true
 	}
 
+	// If any of the main containers have status, then all init containers must
+	// after executed at some point in the past.  However, they could be removed
+	// from the container runtime now, and if we proceed, it would appear as if they
+	// never ran and will re-execute improperly.
+	for i := range pod.Spec.Containers {
```

Comment:
i need to work backwards if this has an impact on pod restarts.

can we add a node e2e that reproduces?

Reply:
we need to not regress the following:

https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#pod-restart-reasons

Reply:
We have always allowed init containers to be rerun.  A key point is we should not rerun them while main status containers are running.  We *must* re run them on reboot.

Reply:
Could add this to only consider main containers if they are running
```diff
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_container.go b/pkg/kubelet/kuberuntime/kuberuntime_container.go
index cb1ad6b306c..48bd0a9ddb2 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_container.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_container.go
@@ -754,6 +754,9 @@ func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus)
                if status == nil {
                        continue
                }
+               if status.State != kubecontainer.ContainerStateRunning {
+                       continue
+               }
                return nil, nil, true
        }
```

Reply:
if the pod sandbox is lost, we guarantee that we rerun init containers today (iirc)

Reply:
@derekwaynecarr also, it would seem that we don't adhere to our own documented behavior already https://github.com/kubernetes/kubernetes/issues/88886

Reply:
> can we add a node e2e that reproduces?

Note to self, the test to verify this case would 1) start a pod with init container 2) while main container is running, call delete container on the init container 3) watch to make sure the init container does not run again.

Not currently sure how to do that or if node e2e can runtime level access.

Reply:
@sjenning the new logic makes sense to me to check primary container status before checking status of any init contianer.